src/hasi/segmentation_pipeline.py

```diff
@@ -0,0 +1,350 @@
+
+    fiducials = []
+    for line in lines:
+        e = line.split(',', 4)
+        p = itk.Point[itk.D, 3]()
+        for i in range(3):
+            p[i] = float(e[i + 1])
+            if ras and i < 2:
+                p[i] = -p[i]
+        fiducials.append(p)
+
+    return fiducials
+
+
+rigid_transform_type = itk.VersorRigid3DTransform[itk.D]
+
+# create an atlas laterality changer transform
+atlas_aa_laterality_inverter = itk.Rigid3DTransform.New()
+invert_superior_inferior = atlas_aa_laterality_inverter.GetParameters()
+# the canonical pose was chosen without regard for proper anatomical orientation
+invert_superior_inferior[8] = -1  # so we mirror along SI axis
+atlas_aa_laterality_inverter.SetParameters(invert_superior_inferior)
+
+
+def register_landmarks(atlas_landmarks, input_landmarks):
+    transform_type = itk.Transform[itk.D, 3, 3]
+    landmark_transformer = itk.LandmarkBasedTransformInitializer[transform_type].New()
+    rigid_transform = rigid_transform_type.New()
+    landmark_transformer.SetFixedLandmarks(atlas_landmarks)
+    landmark_transformer.SetMovingLandmarks(input_landmarks)
+    landmark_transformer.SetTransform(rigid_transform)
+    landmark_transformer.InitializeTransform()
+
+    # force rotation to be around center of femur head
+    rigid_transform.SetCenter(atlas_landmarks[0])
+    # and make sure that the other corresponding point maps to it perfectly
+    rigid_transform.SetTranslation(input_landmarks[0] - atlas_landmarks[0])
+
+    return rigid_transform
+
+
+# If label is non-zero, only the specified label participates
+# in computation of the bounding box.
+# Normally, all non-zero labels contribute to bounding box.
+def label_bounding_box(segmentation, label=0):
+    if label != 0:
+        segmentation = itk.binary_threshold_image_filter(
+            segmentation, lower_threshold=label, upper_threshold=label)
+    image_mask_spatial_object = itk.ImageMaskSpatialObject[3].New(segmentation)
+    bounding_box = image_mask_spatial_object.ComputeMyBoundingBoxInIndexSpace()
+    return bounding_box
+
+
+def process_case(root_dir, bone, case, bone_label, atlas):
+    case_base = root_dir + bone + '/' + case + '-' + atlas  # prefix for case file names
+
+    pose = read_slicer_fiducials(root_dir + bone + '/Pose.fcsv')
+
+    case_landmarks = read_slicer_fiducials(root_dir + bone + '/' + case + '.fcsv')
+    pose_to_case = register_landmarks(case_landmarks, pose)
+
+    if case[-1] != atlas[-1]:  # last letter of file name is either L or R
+        print(f'Changing atlas laterality from {atlas[-1]} to {case[-1]}.')
+        # pose_to_case.Compose(atlas_aa_laterality_inverter, True)
+        pose_to_case.Compose(atlas_aa_laterality_inverter)
+    # we don't need to change laterality of atlas landmarks
+    # as they all lie in a plane with K coordinate of zero
+
+    atlas_bone_label_filename = root_dir + bone + '/' + atlas + '-AA-' + bone + '-label.nrrd'
```

Comment:
Is this filename convention documented and enforced elsewhere in the project as well?

Reply:
It is probably not documented yet. This is just where I have put files on my computer. I guess we should document it in the future, or make it configurable.

Reply:
Overall, we should move away from file IO, file name conventions, to in-memory data structure input output, dagster.